Cleanup server-mocks

uladzimirdev opened the pull request:
Removed the separate endpoint setup for root collection items, included it in the main collection items setup.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/30811)
<!-- Reviewable:end -->

uladzimirdev committed:
Cleanup server-mocks

uladzimirdev commented:
@kulyk @ranquild I'm not sure how to deal with personal collection. Root collection is pretty easy to identify from the collection list, but what about personal, is there any way without explicitly passing `PERSONAL_COLLECTION`?

uladzimirdev commented on frontend/src/metabase/containers/ItemPicker/ItemPicker.unit.spec.js:
```diff
@@ -95,8 +94,7 @@ async function setup({
-  setupCollectionItemsEndpoint({ dashboards: Object.values(DASHBOARD) });
-  setupRootCollectionItemsEndpoint({
+  setupCollectionItemsEndpoint({
     collections: Object.values(COLLECTION).concat(extraCollections),
```
`Object.values().concat` is not only old school, but also a more performant version 😸

uladzimirdev commented on frontend/test/__support__/server-mocks/collection.ts:
```diff
@@ -104,46 +104,44 @@ function setupCollectionWithErrorById({
+
+    // root or personal
+    if (isRootOrPersonal) {
+      if (collectionIdParam === "root") {
```
moved logic here, under the single endpoint setup, any idea how to deal with PERSONAL?

uladzimirdev committed:
Handle personal collection

ranquild commented on frontend/test/__support__/server-mocks/collection.ts:
```diff
@@ -104,46 +104,44 @@ function setupCollectionWithErrorById({
-}
-
 export function setupCollectionItemsEndpoint({
+  collections = [],
```
Could you accept a single collection like search helper? It won't be easy/possible to use this helper with multiple collection anyway - because it can't be the case that the same item is within 2+ collections

ranquild commented on frontend/test/__support__/server-mocks/collection.ts:
```diff
@@ -104,46 +104,44 @@ function setupCollectionWithErrorById({
     const collectionIdParam = url.split("/")[5];
 
     const collectionId = Number(collectionIdParam);
+    const isRootOrPersonal = Number.isNaN(collectionId);
```
If you pass just a single collection you wouldn't need to do this

uladzimirdev commented on frontend/test/__support__/server-mocks/collection.ts:
```diff
@@ -85,7 +85,7 @@ export function setupCollectionByIdEndpoint({
   }
 
-  fetchMock.get(/api\/collection\/\d+/, url => {
+  fetchMock.get(/api\/collection\/\d+|root|personal/, url => {
```
you're right, it's a number

uladzimirdev replied:
I was confused by `type CollectionId = RegularCollectionId | "root" | "personal";`

uladzimirdev replied:
another reason I was confused was in tests

<img width="1254" alt="image" src="https://github.com/metabase/metabase/assets/125459446/77367acc-45cf-4399-a4e4-6a219851a7db">

uladzimirdev replied:
ah, it was just a bug in the picker 😄 
<img width="1820" alt="image" src="https://github.com/metabase/metabase/assets/125459446/c969852b-157b-4955-ada0-42e9fe0f9a6d">

uladzimirdev committed:
Rework and simplify test helper

uladzimirdev committed:
Handle special case in picker item

When fake collection with id "personal" is used, we don't need to make http request, only several personal collections can be listed

uladzimirdev commented:
@kulyk @ranquild I updated test helper, it doesn't accept collections anymore 😄 
Also it turned out we have a special case for "fake" collection with id "personal" which is available only for admins. 
Before last commit we triggered endpoint for a personal collection items: `/api/collection/personal/items`, what was caught by tests (and by @npfitz 😄 separately)

kulyk commented on frontend/test/__support__/server-mocks/collection.ts:
```diff
@@ -85,7 +85,7 @@ export function setupCollectionByIdEndpoint({
   }
 
-  fetchMock.get(/api\/collection\/\d+/, url => {
+  fetchMock.get(/api\/collection\/\d+|root/, url => {
     const collectionIdParam = url.split("/")[5];
     const collectionId = Number(collectionIdParam);
```
Do we need to special case the root collection here? Notice how it's coercing the `collectionIdParam` to a number, and then uses it to find a collection object below. For root collection, it seems like we'll end up with something like that:

```js
const collectionIdParam = "root"
const collectionId = Number(collectionIdParam) // NaN
collections.find(c => c.id === collectionId) // always undefined
```

uladzimirdev committed:
Fixup

kulyk reviewed (approved):
Looks good!
